// iec-61131-3/dist/iec-types.js
"use strict";
/*
https://github.com/jisotalo/iec-61131-3

Copyright (c) 2021 Jussi Isotalo <j.isotalo91@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LINT = exports.LWORD = exports.ULINT = exports.LREAL = exports.REAL = exports.DATE = exports.DATE_AND_TIME = exports.DT = exports.TIME_OF_DAY = exports.TOD = exports.TIME = exports.DWORD = exports.UDINT = exports.DINT = exports.INT = exports.WORD = exports.UINT = exports.SINT = exports.BYTE = exports.USINT = exports.BOOL = exports.WSTRING = exports.STRING = exports.ENUM = exports.ARRAY = exports.UNION = exports.STRUCT = void 0;
const handler = __importStar(require("./iec-type-handler"));
/**
 * IEC 61131-3 type: STRUCT - Handles STRUCT data type, provide struct children as object
 *
 * @param children Children variables as IEC object, like: `{intVal: INT, boolVal: BOOL, structVal: STRUCT({...})}`
 * @returns IecType object
 */
const STRUCT = (children) => new handler.STRUCT(children);
exports.STRUCT = STRUCT;
/**
 * IEC 61131-3 type: UNION - Handles UNION data type, provide union children as object
 *
 * @param children Children variables as IEC object, like: `{intVal: INT, boolVal: BOOL, structVal: STRUCT({...})}`
 * @returns IecType object
 */
const UNION = (children) => new handler.UNION(children);
exports.UNION = UNION;
/**
 * IEC 61131-3 type: ARRAY - Handles 1..3 dimensional arrays.
 *
 * Example with 1-dimensional INT array of 10 values: `ARRAY(INT, 10)`

 * Example with 2-dimensional REAL array of 2*5 values: `ARRAY(REAL, [2, 5])`
 *
 * @param dataType Data type of the array (example: INT)
 * @param dimensions Array dimension as number (if 1-dimensional array). If multi-dimensional, array dimensions as array of numbers (like `[2, 5]`)
 */
const ARRAY = (dataType, dimensions) => new handler.ARRAY(dataType, dimensions);
exports.ARRAY = ARRAY;
/**
 * IEC 61131-3 type: ENUM
 * Handles enumeration types with different data types
 *
 * @param definition Enumeration definition as object (like `{key1: 1, key2: 2}`)
 * @param dataType Data type of the ENUM (default is iec.INT)
 * @returns
 */
const ENUM = (definition, dataType) => new handler.ENUM(definition, dataType);
exports.ENUM = ENUM;
/**
 * IEC 61131-3 type: STRING - Default length 80 characters
 * @param length Length of the string variable (similar as in the PLC), default is 80
 */
const STRING = (length) => new handler.STRING(length);
exports.STRING = STRING;
/**
 * IEC 61131-3 type: WSTRING - Default length 80 characters
 * @param length Length of the string variable (similar as in the PLC), default is 80
 */
const WSTRING = (length) => new handler.WSTRING(length);
exports.WSTRING = WSTRING;
/**
 * IEC 61131-3 type: BOOL (1 byte)
 */
exports.BOOL = new handler.BOOL();
/**
 * IEC 61131-3 type: USINT (1 byte)
 */
exports.USINT = new handler.USINT();
/**
 * IEC 61131-3 type: BYTE (1 byte)
 */
exports.BYTE = new handler.BYTE();
/**
 * IEC 61131-3 type: SINT (1 byte)
 */
exports.SINT = new handler.SINT();
/**
 * IEC 61131-3 type: UINT (2 bytes)
 */
exports.UINT = new handler.UINT();
/**
 * IEC 61131-3 type: WORD (2 bytes)
 */
exports.WORD = new handler.WORD();
/**
 * IEC 61131-3 type: INT (2 bytes)
 */
exports.INT = new handler.INT();
/**
 * IEC 61131-3 type: DINT (4 bytes)
 */
exports.DINT = new handler.DINT();
/**
 * IEC 61131-3 type: UDINT (4 bytes)
 */
exports.UDINT = new handler.UDINT();
/**
 * IEC 61131-3 type: DWORD (4 bytes)
 */
exports.DWORD = new handler.DWORD();
/**
 * IEC 61131-3 type: TIME (4 bytes)
 */
exports.TIME = new handler.TIME();
/**
 * IEC 61131-3 type: TOD (4 bytes)
 */
exports.TOD = new handler.TOD();
/**
 * IEC 61131-3 type: TIME_OF_DAY (4 bytes)
 */
exports.TIME_OF_DAY = new handler.TIME_OF_DAY();
/**
 * IEC 61131-3 type: DT (4 bytes)
 * TODO: Conversion to Javascript Date object?
 */
exports.DT = new handler.DT();
/**
 * IEC 61131-3 type: DATE_AND_TIME (4 bytes)
 * TODO: Conversion to Javascript Date object?
 */
exports.DATE_AND_TIME = new handler.DATE_AND_TIME();
/**
 * IEC 61131-3 type: DATE (4 bytes)
 * TODO: Conversion to Javascript Date object?
 */
exports.DATE = new handler.DATE();
/**
 * IEC 61131-3 type: REAL (4 bytes)
 */
exports.REAL = new handler.REAL();
/**
 * IEC 61131-3 type: LREAL (4 bytes)
 */
exports.LREAL = new handler.LREAL();
/**
 * IEC 61131-3 type: ULINT (8 bytes)
 * TODO: Requires Node.js that supports BigInt
 */
exports.ULINT = new handler.ULINT();
/**
 * IEC 61131-3 type: LWORD (8 bytes)
 * TODO: Requires Node.js that supports BigInt
 */
exports.LWORD = new handler.LWORD();
/**
 * IEC 61131-3 type: LINT (8 bytes)
 * TODO: Requires Node.js that supports BigInt
 */
exports.LINT = new handler.LINT();
